EqualExpectation failure messages for unequal strings

aarongodin opened the pull request:
Addresses https://github.com/crystal-lang/crystal/issues/4432. Failure messages for checking equality of two strings will output the string using a percent-delimiter if the string contains a quote character.

@straight-shoota Mentioned on the issue the option of not having delimiters in the first place, but I think it makes it more apparent to the user what is being shown. That said, we don't show delimiters or type indication for other primitives, unless it's ambiguous to the error message what the type difference is between the two values, so maybe I'm in the minority on that preference. Showing the delimiter is better to me since you can copy the inspected value as a valid expression in crystal.

This is my first PR, so please let me know of any feedback or additional considerations that I missed.

straight-shoota reviewed (commented):
This looks great overall!

A few proposals for improvements:
1. If the string contains an unmatched end delimiter, it needs to be escaped.
2. I would suggest using distinct opening and closing delimiters for the percent literals. They're easier to read. Parenthesis (`%()`) is most common.

I also have some more advanced ideas, but maybe you feel up to implementing them. I'll happily accept this PR without them, though.

3. Would it perhaps make sense to use the same delimiters for expected and actual value? `"foo".should eq "bar\""` would use double quotes and percent literal delimiters, respectively. Using the same style for both would make it easier to compare them.
4. Combining 1. and 2., delimiters could even be chosen based on the string contents to reduce the need for escaping.

aarongodin commented:
> 1. If the string contains an unmatched end delimiter, it needs to be escaped.

Makes sense, definitely will fix.

> 3. Would it perhaps make sense to use the same delimiters for expected and actual value? `"foo".should eq "bar\""` would use double quotes and percent literal delimiters, respectively. Using the same style for both would make it easier to compare them.

I agree, that would improve readability. I'll implement this as well.

> 4. Combining 1. and 2., delimiters could even be chosen based on the string contents to reduce the need for escaping.

I like this idea. I started down a solution that chooses the least-common delimiter in the two strings and uses that. However, I think that solution would conflict with preferring `%(`. Is there a reason `%(` is preferred, or is it just the most common?

Thanks for the help!

straight-shoota commented:
> However, I think that solution would conflict with preferring `%(`. Is there a reason `%(` is preferred, or is it just the most common?

Not really. Or rather, if the string contains unmatched `)`, I think it's fine to prefer a different delimiter that works better (i.e. without escaping).

Not sure if there's a particular reason for `%(`, but it's prevalent in both Ruby and Crystal.

aarongodin commented:
I went ahead and updated the implementation to favor parens, but fall back to `#inspect` if there are simply more parens than quotes. Thinking it through, I'm wondering how often that would happen and whether it's worth the extra effort. Or do we simply use parens and call it a day?

I did make it so that the two strings will both use the same formatting (either `%()` or `#inspect`).

To perform the escaping of the unbalanced delimiters, I added a method to `String`. It seemed like something that could be generally useful, but I would certainly defer to folks on the core team on whether it belongs there. It can be a method on `EqualExpectation` if it doesn't seem generically useful.

Thanks again for the help!
